Fix inductor `end` of slice.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106208
* #105923

This PR fixes checks for the `end` integer when slicing a tensor. The problem was that
slicing dimension `dim` with `(start, end, step)` created (among others) the guard: `end <
size[dim]`.

Although, that doesn't generate incorrect results, a tighter guard would be: `end <
size[dim] + 1`, since `end` is not included in the slice.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ysiraichi committed:
Fix inductor `end` of slice.

This PR fixes checks for the `end` integer when slicing a tensor. The problem was that
slicing dimension `dim` with `(start, end, step)` created (among others) the guard: `end <
size[dim]`.

Although, that doesn't generate incorrect results, a tighter guard would be: `end <
size[dim] + 1`, since `end` is not included in the slice.

[ghstack-poisoned]

ysiraichi committed:
Update on "Fix inductor `end` of slice."


This PR fixes checks for the `end` integer when slicing a tensor. The problem was that
slicing dimension `dim` with `(start, end, step)` created (among others) the guard: `end <
size[dim]`.

Although, that doesn't generate incorrect results, a tighter guard would be: `end <
size[dim] + 1`, since `end` is not included in the slice.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Fix inductor `end` of slice."


This PR fixes checks for the `end` integer when slicing a tensor. The problem was that
slicing dimension `dim` with `(start, end, step)` created (among others) the guard: `end <
size[dim]`.

Although, that doesn't generate incorrect results, a tighter guard would be: `end <
size[dim] + 1`, since `end` is not included in the slice.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented:
You can't do this, and worse, I don't think there is any easy way to solve this problem.

The problem is that the contiguity between `x[:, :end]` and `x[:, :end-1]` is different. The former is contiguous (since you didn't actually do anything); the latter is discontiguous, as you skip the last element each row. Thus, there must be a guard to distinguish these two cases.

BTW, that's what https://github.com/pytorch/pytorch/pull/105535 is for, to make it so we don't need to guard here.

ysiraichi committed:
Update on "Fix inductor `end` of slice."


This PR fixes checks for the `end` integer when slicing a tensor. The problem was that
slicing dimension `dim` with `(start, end, step)` created (among others) the guard: `end <
size[dim]`.

Although, that doesn't generate incorrect results, a tighter guard would be: `end <
size[dim] + 1`, since `end` is not included in the slice.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi commented:
Ah, I see. I will close this PR.
